README.md
# django painless redirects

[![CI](https://img.shields.io/github/workflow/status/bnzk/django-painless-redirects/CI.svg?style=flat-square&logo=github "CI")](https://github.com/bnzk/django-painless-redirects/actions/workflows/ci.yml)
[![Version](https://img.shields.io/pypi/v/django-painless-redirects.svg?style=flat-square "Version")](https://pypi.python.org/pypi/django-painless-redirects/)
[![Licence](https://img.shields.io/github/license/bnzk/django-painless-redirects.svg?style=flat-square "Licence")](https://pypi.python.org/pypi/django-painless-redirects/)
[![PyPI Downloads](https://img.shields.io/pypi/dm/django-painless-redirects?style=flat-square "PyPi Downloads")](https://pypistats.org/packages/django-painless-redirects)

like django.contrib.redirects, on steroids. maybe.


## Features

- simple redirects table, with that bit more flexibility / convenience
    - django sites framework support
    - wildcard matching
    - explicit limit to site/domain
    - explicit redirect to site
    - decide if you want to keep GET vars
    - move complete trees (/old/productXY to /new/productXY, with one redirect, for all products)
- force site domain middleware, that redirects to current site's domain, if not already there

Yet to be done:

- decide if to include GET vars when matching a redirect
- contrib packages with "magic" redirects for django-cms, django-filer -> SEO getting easy.


## Installation

To get the latest stable release from PyPi

    pip install django-painless-redirects

Add ``painless_redirects`` to your ``INSTALLED_APPS``

    INSTALLED_APPS = (
        ...,
        'painless_redirects',
    )

Add the following middlware to MIDDLEWARE, to make basic manual redirects work. 

    painless_redirects.middleware.ManualRedirectMiddleware

As a rule of thumb, you'll add the middleware rather at the end of the list. If
you add it more at the beginning, other redirecting middlewares will 
get priority - for example, you'll not be able to redirect any path 
without having a language slug first, when LocaleMiddleware is before 
our middleware.

If you want to always redirect to the domain name entered in your current site (django.contrib.sites must be installed),
also add this middleware (position is not that important):

    painless_redirects.middleware.ForceSiteDomainRedirectMiddleware


## Compatibility

- django 1.11: should use `django-painless-redirects<0.3.6` (not supported anyway since april 2020)


## Usage

Add and manage your redirects in the django admin panel, under "painless redirects" > "redirects". Check hits, 
check auto created redirects, enable them if needed.


## Settings

- `PAINLESS_REDIRECTS_INDEXED_CHARFIELD_MAX_LENGTH (default: 800)` - needed for mysql (set to ~255). forces you to add `'painless_redirects'` to `settings.MIGRATION_MODULES`
- `PAINLESS_REDIRECTS_AUTO_CREATE (default: True)` - auto create redirects when a 404 is encountered
- `PAINLESS_REDIRECTS_AUTO_CREATE_ENABLED (default: False)` - should auto created redirects be enabled instantly? beware, enabling this feature might ruin your SEO, but could also be a real timesaver...
- `PAINLESS_REDIRECTS_AUTO_CREATE_TO_PATH (default: '/')` - where to redirect to, when a 404 is auto created
- `PAINLESS_REDIRECTS_AUTO_CREATE_SITE (default: True)` - limit auto created redirects to the current site (from the django sites framework)


## Development

- there is test app, available with `./manage.py runserver`.
- quick tests, within your current env: ./manage.py test
- to run tests with django 1.11 / 2.0 / 2.1 / ... : `tox`


## Contributions

If you want to contribute to this project, please perform the following steps

.. code-block:: bash

    # Fork this repository
    # Clone your fork
    mkvirtualenv django-painless-redirects
    pip install -r test_requirements.txt
    git checkout -b feature_branch
    # Implement your feature and tests
    git add . && git commit
    git push -u origin feature_branch
    # Send us a pull request for your feature branch


CHANGELOG.txt
=== (ongoing) ===
- contrib packages that add redirects automagically


=== 0.7.2 (wip) ===
-


=== 0.7.1 (2022-06-14) ===
- badges revamped
- README.md for PyPi
- Django 4.0 compatibility established
- Python 3.10 in test matrix


=== 0.7.0 (2022-03-16) ===
- compatible with django 3.1+


=== 0.6.0 (2021-09-28) ===
- #25 changelist with more djangolistic links, still not satisfied
- #24 changelist editable. yea.


=== 0.5.0 (2021-09-28) ===
- compatibility for django 3+
- basic admin tests, changelist and detail/change view


=== 0.4.2 (2021-02-24) ===
- changed CI from travis to github actions
- race condition bug #22 (#21)
- managemenet command to eliminiate duplicate redirecthits (that would violate the new `unique_together` constraint)


=== 0.4.1 (2021-01-18) ===
- configure mysql max_length problem for old_path/referer!
- better support and test multilingual middleware (#19)
- better testing for wildcard matches with site enabled (right_path bug!)
- docs updated (middleware order)


=== 0.4.0 (2020-11-30) ===
- not a breaking, but loosing content change: hit counting changed (all hits will be reset to 0!)
- hit counting changed, now by referer, for a better overview
- bug fixes.


=== 0.3.7 (2020-07-03) ===
- integrity error fix (when auto creation was enabled)


=== 0.3.6 (2020-06-22) ===
- prepare referrer counting
- docs
- bug fix: when existing auto created exact matchin but not enabled redirect was there,
  it prevented use of enabled matchting wildcard redirect
- using django 2.2 for creation of new migrations - dont use django 1.11 with this release and up.


=== 0.3.5 (2020-04-20) ===
- fix migrations chaos in wheel version, finally :-)


=== 0.3.4 (2020-04-20) ===
- fix migrations chaos in wheel version!


=== 0.3.3 (2020-04-20) ===
- mysql index fix for max_length. old_path can now be 800 chars max


=== 0.3.2 (2020-04-20) ===
- no more exceptions when redirects with too long urls are auto created
- admin actions for deleting auto created redirects (all/disabled only)


=== 0.3.1 (2019-12-22) ===
- resolving mess with PYPI.rst


=== 0.3.0 (2019-12-22) ===
- NOT ON PYPI!
- auto create redirects on 404 hits
- log/count 404 hits
- settings for log and auto creation
- drop py2!
- officialy support django 1.11 up to 3.0


=== 0.2.1 (2018-11-07) ===
- adding a missing migration (updating labels!)


=== 0.2 (2018-08-30) ===
- new feature: keep treee
- new feature: keep getvars
- new feature: 301 OR 302 redirects
- new feature: MIDDLEWARE setting compatible (django 1.10+ middleware style)
- should be backward compatible
- should work up to django 2.1
- support for django below 1.11: still working in tests, but, yea, you should update!


=== 0.1.5 ===
- fixing broken admin regression from 0.1.4


=== 0.1.4 ===
- bug when specifying the "to" site
- translations
- cleanup


=== 0.1.3 ===
- added missing migration


=== 0.1.2 ===
- last try with pypi's readme


=== 0.1.1 ===
- ui changes in admin change view
- pypi mess and readme updated


=== 0.1 ===
- intitial release 0.1
- basics there


# Suggested file syntax:
#
# === (ongoing) ===
# - this is always on top of the file
# - when you release a new version, you rename the last `(ongoing)` to the new
#   version and add a new `=== (ongoing) ===` to the top of the file
#
# === 1.0 ===
# - a major version is created when the software reached a milestone and is 
#   feature complete
#
# === 0.2 ===
# - a minor version is created when new features or significant changes have
#   been made to the software.
#
# === 0.1.1 ==
# - for bugfix releases, fixing typos in the docs, restructuring things, simply
#   anything that doesn't really change the behaviour of the software you
#   might use the third digit which is also sometimes called the build number.


painless_redirects/models.py
from django.db import models
from django.contrib.sites.models import Site
from django.db.models import Sum
from django.template.defaultfilters import truncatechars
from django.urls import reverse
from django.utils.safestring import mark_safe
from django.utils.html import format_html
from django.utils.translation import gettext_lazy as _

from painless_redirects import conf

try:
    from django.utils.encoding import python_2_unicode_compatible
except ImportError:
    def python_2_unicode_compatible(f):
        return f

from painless_redirects.managers import RedirectQuerySet

REDIRECT_TYPE_CHOICES = (
    (301, "Moved permanently"),
    (302, "Moved temporary"),
)


@python_2_unicode_compatible
class RedirectHit(models.Model):
    redirect = models.ForeignKey(
        'painless_redirects.Redirect',
        on_delete=models.CASCADE,
    )
    referer = models.CharField(
        _(u'Referer'),
        # check https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers
        max_length=conf.INDEXED_CHARFIELD_MAX_LENGTH,
        help_text=_("Where the hit comes from (ie refering url")
    )
    hits = models.PositiveIntegerField(
        default=0,
        editable=False,
    )
    last_hit = models.DateTimeField(
        blank=True,
        null=True,
        auto_now=True,
        help_text='The last time the URL/Redirect was hit, from this referer.'
    )

    class Meta:
        ordering = ('-hits',)
        unique_together = ('redirect', 'referer')

    def __str__(self):
        return '{}x from {}'.format(self.hits, self.referer)


@python_2_unicode_compatible
class Redirect(models.Model):
    enabled = models.BooleanField(
        default=True,
        verbose_name=_(u'Enabled'),
        help_text=_("Shall this redirect be effectivly used?"),
    )
    ignored = models.BooleanField(
        default=False,
        verbose_name=_(u'Ignored'),
        help_text=_("Shall this redirect be ignored? (use to tighen/cleanup your redirects list)"),
    )
    auto_created = models.BooleanField(
        default=False,
        verbose_name=_(u'Auto created'),
        help_text=_("Created by a 404 hit? (must be enabled via settings)"),
        editable=False,
    )
    permanent = models.BooleanField(
        default=True,
        verbose_name=_(u'Permanent'),
        help_text=_("Use status code 301. For temporary fixes, uncheck to use a status code of 302"),
    )
    # max length note: mysql index cannot be more than 307Xbytes, so with utf-8,
    # we can have no more than max_length around 800 for old_path
    old_path = models.CharField(
        _(u'From path'),
        # check https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers
        max_length=conf.INDEXED_CHARFIELD_MAX_LENGTH,
        db_index=True,
        help_text=_("Absolute path, excluding the domain name. Example: '/events/search/'")
    )
    wildcard_match = models.BooleanField(
        default=False,
        verbose_name=_(u'Wildcard match'),
        help_text=_('Add wildcard (/from/path/*)'),
    )
    # too complicated to implement for now!
    # querystring_match = models.BooleanField(
    #     default=False,
    #     verbose_name=_(u'Querystring match'),
    #     help_text=_('Considers ?search=whatever when mathing)'),
    # )total_hits
    site = models.ForeignKey(
        Site,
        null=True,
        on_delete=models.CASCADE,
        blank=True,
        related_name="redirect_old_site",
        help_text=_('Optional, limit redirect to this site'),
    )
    domain = models.CharField(
        max_length=64,
        blank=True,
        default='',
        help_text=_('Optional, exlicitly limit to specific domain'),
    )
    new_path = models.CharField(
        _(u'To path'),
        max_length=255,
        help_text=_('Absolute path, or full url (with http://.../)'),
    )
    keep_tree = models.BooleanField(
        default=False,
        verbose_name=_("Keep tree"),
        help_text=_("Only applies when wildcard matching enabled"),
    )
    keep_querystring = models.BooleanField(
        default=False,
        verbose_name=_("Keep querystring"),
        help_text=_("Re-applies GET querystring, if any (?page=4&search=banana)"),
    )
    new_site = models.ForeignKey(
        Site,
        null=True,
        on_delete=models.CASCADE,
        blank=True,
        related_name="redirect_new_site",
        help_text=_('Optional, automatically insert correct domain name of this site'),
    )

    objects = RedirectQuerySet.as_manager()

    class Meta:
        verbose_name = "Redirect"
        verbose_name_plural = "Redirects"
        unique_together = (('site', 'domain', 'old_path', ), )
        ordering = ('old_path', )

    def redirect_value(self, scheme, right_path='', querystring=''):
        if self.new_site:
            url = "%s://%s%s" % (scheme, self.new_site.domain, self.new_path)
        else:
            url = self.new_path
        if right_path and self.keep_tree:
            url = '{}{}'.format(url, right_path)
        if querystring and self.keep_querystring:
            url = '{}?{}'.format(url, querystring)
        return url

    def total_hits(self):
        qs = self.redirecthit_set.all().aggregate(total_hits=Sum('hits'))
        return qs['total_hits']
    total_hits.admin_order_field = 'total_hits'

    def old_loc_display_data(self):
        path = self.old_path
        if self.domain:
            loc = "%s%s" % (
                self.domain, path
            )
            loc_link = loc
            if not self.domain.startswith('http'):
                loc_link = 'http://{}'.format(loc)
        else:
            loc = "%s%s" % (
                getattr(self.site, "domain", ""), path
            )
            loc_link = loc
            if getattr(self.site, "domain", None):
                loc_link = 'http://{}'.format(loc)
        if self.wildcard_match:
            loc += "*"
        return truncatechars(loc, 50), loc_link

    def old_loc(self):
        loc, loc_link = self.old_loc_display_data()
        loc = format_html(
            ' <a href="{}">{}</a> (<a href="{}" target="_blank">check</a>)',
            reverse('admin:painless_redirects_redirect_change', args=(self.id, ), ),
            loc,
            loc_link,
        )
        return mark_safe(loc)
    old_loc.short_description = 'From'

    def new_loc_display_data(self):
        loc = self.redirect_value('http')
        loc_link = loc
        # redirect_value has http covered
        # if getattr(self.new_site, "domain", None):
        #     loc_link = 'http://{}'.format(loc)
        if self.keep_tree:
            loc += '/' if not loc.endswith('/') else ''
            loc += "keep-wildcard/"
        if self.keep_querystring:
            loc += "?keep=query"
        return loc, loc_link

    def new_loc(self):
        loc, loc_link = self.new_loc_display_data()
        loc = format_html(
            '<a href="{}" target="_blank">{}</a>',
            loc_link,
            loc,
        )
        return mark_safe(loc)
    new_loc.short_description = 'To'

    def __str__(self):
        old_loc, old_loc_link = self.old_loc_display_data()
        new_loc, new_loc_link = self.new_loc_display_data()
        return '{} ---> {}'.format(old_loc, new_loc)
